Line comment bug fix

heysupratim opened the pull request:
Better line comment toggling , initiates the line comment at the start of the code and maintains the position of the cursor throughout (like sublime )

Better for lint to comment indents 

Affected me so hence changed it 

Will fail some editorCommandHandler tests which expect the line comment at the start of the line

fix for https://github.com/adobe/brackets/issues/11953#issuecomment-159572261

heysupratim committed:
Added line comment toggling only at start of code added #11953

heysupratim committed:
added line toggling like sublime where subsequent comment position is based on first toggle position fixed #11953

abose commented:
@borax12 Thanks for contributing to brackets.
There is a slight issue here
![linecommentbug](https://cloud.githubusercontent.com/assets/5336369/11395838/7d6e0a1a-9393-11e5-95d9-d7d53c80661a.gif)
When you include a selection with uneven cursor position, it tries to comment out in the wrong place, and toggle comment breaks.
This could be fixed if you analyze the left most line position in the selection and then use that position for inserting the comments.

borax12 commented:
Ah , i see, checking it out

heysupratim committed:
uneven seelcted led to wrong toggle of comments , recalucated curssor position for offset lines #11953

heysupratim committed:
lint error of equality - lint is awesome

heysupratim committed:
removed duplicate variable declaration

borax12 commented:
@abose i had to push extra commits due to some code formatting issues . Hope it works now

abose commented:
unit tests for `editor command handlers` fails after this change. The unit tests needs to be modified with the new commenting structure.

heysupratim committed:
re wrote line comment test cases

heysupratim committed:
removed lint errors in EditorCommandHandler-test.js

borax12 commented:
@abose - Rewrote the test cases , The editor command handler Unit Test Cases Pass now

redmunds commented:
It looks like you're changing it so that inline comments are *always* indented. Is that correct?

This needs to be a preference because some languages (VBScript, PHP, bash?) require that inline comment delimiter is in the first column. Then you'll need tests for both cases.

borax12 commented:
@redmunds yeah it is always indented . I checked the behaviour in ST2 for php and bash scripts - they are also always indented .

Does brackets have per language based preferences ?

abose commented:
:cool: Verified, all unit tests passing now.
You could have a preference like 
"dontIntentComments" {"languages":['bash']} ..
With reasonable defaultpreferences set for languages.

redmunds commented:
> You could have a preference like `"dontIntentComments" {"languages":['bash']}`

Brackets already has a mechanism for specifying language specific preferences, so no need to invent a new one. Take a look at [this example](https://github.com/adobe/brackets/wiki/How-to-Use-Brackets#a-sample-preferences-json-file-) in the How to Use Brackets doc, especially the "language" section:

```json
    "language": {
        "javascript": {
            ...
        }
    },
```

As far as what to call the new preference, it depends on what the default is. Since the current default is not to indent, then a name like `indentInlineComments` would be consistent with existing behavior since prefs that are not defined, evaluate to "falsy". If we want to change the default to be indented, then maybe `dontIndentInlineComments`.

borax12 commented:
@redmunds @abose - Ohk , will take a bit of time as i need to see how preference loading and initiating works in brackets . but i will have to research which languages take line comment indentation by default

TomMalbran commented:
If some languages require that the comments start on the first column, it should be a property of the language definition and not a preference. There can be a preference if an user decides to overwrite such behaviour or wants to go back to have comments at the start of a line.

abose commented:
Never knew that language based pref existed :sweat_smile: Thanks @redmunds @TomMalbran

redmunds commented:
@borax12 There is already PR https://github.com/adobe/brackets/pull/11574 that looks like it fixes the same issue.

@MarcelGerber What's the status of your PR?

abose commented:
@borax12 Could you also look at #11574 and see if Your implementation has any functionality difference with it.
#11574 also doesn't seem to implement any unit tests.

borax12 commented:
i will do that @arun . I have been trying to get the per language
based setting correctly . I will also look at #11574

On 12/1/15, Arun Bose <notifications@github.com> wrote:
> @borax12 Could you also look at #11574 and see if Your implementation has
> any functionality difference with it.
> #11574 also doesn't seem to implement any unit tests.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/adobe/brackets/pull/11954#issuecomment-160886902


-- 

[image: Supratim Chakraborty on about.me]

Supratim Chakraborty
about.me/borax12
  <http://about.me/borax12>

abose commented:
@borax12 Did you get any chance to look into the isues? This is a very good usability improvement that we could ship as part of 1.6 coming late december or early jan.

heysupratim committed:
Merge remote-tracking branch 'upstream/master' into lineCommentBugFix

heysupratim committed:
added preference for Editor linecommentIndent

heysupratim committed:
added preference handling -preference indentLineComment with default value false for fix #11953

heysupratim committed:
made test cases for both prefernce set values fix #11953

heysupratim committed:
removed merge conflicts and added comments

borax12 commented:
@abose  @redmunds i added the default preference and also added tests to accompany both cases for this preference being true or false

borax12 commented:
@redmunds - i added to the defaultPreference

borax12 commented:
@abose - status update for this ?

borax12 commented:
@redmunds @abose - thread bump

petetnt commented on src/editor/EditorCommandHandlers.js:
```diff
@@ -206,8 +207,30 @@ define(function (require, exports, module) {
+                //check if preference for indent line comment is available otherwise go back to default indentation
+                if (Editor.getIndentLineComment()) {
+                    //ignore the first line and recalculate cursor position for first non white space char of every line
+                    if (i!==startLine) {
```
Nitpick: spaces around operators `(i !== startLine)`

petetnt commented on src/editor/EditorCommandHandlers.js:
```diff
@@ -206,8 +207,30 @@ define(function (require, exports, module) {
+                        firstCharPosition = line.search(/\S|$/);
+                    }
+                    //if the non space first character position is before original start position , put comment at the new position otherwise older pos
+                    if (firstCharPosition<originalCursorPosition) {
```
Nitpick: Spaces around operators `(firstCharPosition < originalCursorPosition)`, `cursorPosition = firstCharPosition`, `cursorPosition = originalCursorPosition`

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -171,8 +174,7 @@ define(function (require, exports, module) {
         /**
-         * Invokes Toggle Line or Block Comment, expects the given selection/cursor & document text, invokes
-         * it a 2nd time, and then expects the original selection/cursor & document text again.
+         * Invokes Toggle Line or Block Comment, expects the given selection/cursor & document text,
```
Nitpick: extra `,` at the end of the comment

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -201,12 +200,10 @@ define(function (require, exports, module) {
-                expectSel(startingSel);
-            });
+            
+            //removed invoking the toggling for 2nd time to get original selection back 
```
The comment is not necessary in the code, should be on the PR instead

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -221,105 +218,213 @@ define(function (require, exports, module) {
 
-                testToggleLine(expectedText, {start: {line: 1, ch: 6}, end: {line: 1, ch: 14}});
+                    testToggleLine(expectedText, {start: {line: 1, ch: 6}, end: {line: 1, ch: 14}});
+                }else{
```
Nitpick: spaces around `else`

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -35,6 +36,8 @@ define(function (require, exports, module) {
 
     require("editor/EditorCommandHandlers");
+    
+    var shouldIndentLineComment = Editor.getIndentLineComment();
```
The tests should probably run the tests both for `shouldIndentLineComment` `false` and `true`, not by taking the current value from the prefs :fire:

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -362,41 +468,81 @@ define(function (require, exports, module) {
-                var startingContent = lines.join("\n");
-                myDocument.setText(startingContent);
+                
+                if(shouldIndentLineComment){
```
Nitpick: spaces `if (shouldIndentLineComment) {`, similarly to others.

petetnt commented:
Made some comments, mostly just nitpicks on styles (spacing), but the tests should most likely be written for both preference cases (true/false) separately.

borax12 commented:
Hey @petetnt , will be going through the styling issues , i thought i did adhere to all lint checks but yeah there might be some conventions i made mistakes about , regarding the test, because its a user dictated feature , i wrote test cases which cover both based on the toggling of the feature. Lot of boilerplate :(

petetnt commented:
I don't think the linter enforces spacing right now so it's mostly just to fit the other code :) 

The boilerplate kinda sucks, but that way you can ensure that both cases are being tested (you could try flipping the pref on `afterEach` and running the test again but not sure if that's possible.

borax12 commented:
Well i will be working on the spacing and forced execution of both tests . Expect this to drop down very soon good sir :)

heysupratim committed:
Merge branch 'master' into lineCommentBugFix

heysupratim committed:
Removed styling issues and removed unwanted comments

Removed inovking the toggling of comment for 2nd time to get original selection back as the line commend indent feature doesnt
give the same bheviour as an editor without line comment indentation

borax12 commented:
@petetnt i did work out all the styling issues , but still havent done the seperation of tests for both cases of preference , not able to find the correct way to set and and reset the preference for testing both cases of preference values

petetnt commented:
I think the best way is just to set the preference to true when starting, then creating test `should change the preference` which flips the switch and then do the another set of tests below `should NOT indent when foo is set to false` and so on.

heysupratim committed:
Merge branch 'master' into lineCommentBugFix

heysupratim committed:
seperated the test cases for both preference set and unset and reset preference back to original value

borax12 commented:
@petetnt - i separated the test cases with one set making it run with preference set to true and other making it run on preference set to false

borax12 commented:
@petetnt - is there anything else needed on this , just needed an update :)

petetnt commented:
Sorry @borax12, I have been swamped with work so I haven't had the time to review the latest commit. I'll do it tomorrow morning if that is okay with you :+1:

borax12 commented:
no worries there @petetnt - appreciate the update .And regarding the work . I hope it turns out the best for you :)

borax12 commented:
@petetnt - updates? :)

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -217,30 +211,45 @@ define(function (require, exports, module) {
-            beforeEach(setupFullEditor);
+        describe("Line comment/uncomment with indentLineComment enabled", function () {
+
+            beforeEach(function () {
```
Style nit: Indentation with the `beforeEach` and `afterEach`

borax12 replied:
what exactly needs to be indented here ?

petetnt replied:
Sorry about being vague: I mean that there's extra indentation within this function block:
```
        beforeEach(function () {
                setupFullEditor();
                PreferencesManager.set("indentLineComment",true);
            });
```

Should be:

            beforeEach(function () {
                setupFullEditor();
                PreferencesManager.set("indentLineComment",true);
            });

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -466,9 +489,314 @@ define(function (require, exports, module) {
+
+        describe("Line comment/uncomment with indentLineComment disabled", function () {
+
+            beforeEach(function () {
```
Style nit: Indentation with the `beforeEach` and `afterEach`

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -466,9 +489,314 @@ define(function (require, exports, module) {
+
+                it("should toggle comments on separate lines with cursor selections", function () {
+
+                    myEditor.setSelections([{start: {line: 1, ch: 4}, end: {line: 1, ch: 4}},
```
Stylenit: the `{start:`s should be aligned (check the other changes to `setSelections` too

borax12 replied:
@petetnt - again , could you give a pointer as to how its done , the tests are written like this in master since day 1 so have not got any idea about this particular style nit

petetnt replied:
Again sorry about being vague: I meant this and the next line:

Is: 
```
myEditor.setSelections([{start: {line: 1, ch: 4} ...
                     {start: {line: 1, ch: 4} ...
```

Should be:

```
myEditor.setSelections([{start: {line: 1, ch: 4} ...
                        {start: {line: 1, ch: 4} ...
```

(These are really minor nits so feel free to disregard)

borax12 replied:
@petetnt - i am disregarding this stylenit  , is that ok , did the other changes and currently splitting the last remaining test

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -1836,21 +2173,39 @@ define(function (require, exports, module) {
 
             it("should line comment/uncomment generic JS code", function () {
-                myEditor.setCursorPos(10, 0);
+                if (shouldIndentLineComment) {
```
Should most likely be separated like the other tests

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -1902,18 +2257,33 @@ define(function (require, exports, module) {
 
                 it("should handle multiple selections in different regions, toggling line selection (but falling back to block selection in HTML/CSS)", function () {
-                    myEditor.setSelections([{start: {line: 1, ch: 4}, end: {line: 1, ch: 10}},
+                    if (shouldIndentLineComment) {
```
Should be separated to with / without `shouldIndentLineComments` like the other tests

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -466,9 +489,314 @@ define(function (require, exports, module) {
+                    lines[5] = "    }";
+                    var expectedText = lines.join("\n");
+
+                    testToggleLine(expectedText, [{ start : { line : 1, ch : 4 }, end : { line : 1, ch : 4 }, reversed : false, primary : false },
```
Stylenit: the test object has different spacing than the other `testToggleLine`'s (no space before, space after `,` and `;`

petetnt commented on test/spec/EditorCommandHandlers-test.js:
```diff
@@ -466,9 +489,314 @@ define(function (require, exports, module) {
+                    lines[3] = "        //a();";
+                    var expectedText = lines.join("\n");
+
+                    testToggleLine(expectedText, [{ start: {line: 1, ch: 6 }, end: {line: 1, ch: 8 }, reversed: false, primary: false },
```
Stylenit: the test object has different spacing than the other `testToggleLine`'s (no space before, space after `,` and `;`

petetnt commented:
@borax12 Sorry for taking so long :snail: 

I made some more comments, mostly just style nitpicks but there was a single test that wasn't changed to split format later in the file too.

heysupratim committed:
Merge branch 'master' into lineCommentBugFix

borax12 commented:
@petetnt - i made the final changes , split the two tests and tried to address the top mentioned style nits , hopefully this should do it :)

heysupratim committed:
split the remaining two tests , hopefully this ends the brackets linecommentindent feature- that was a tough one month

borax12 commented:
@petetnt - hey man , i must be bugging you , but i did split the changes, i guess that should do it

petetnt commented:
LGTM /cc @abose

abose commented:
Thanks @borax12 @petetnt  .Will check this out over the weekend .

abose commented:
LGTM. Verified all the test cases hilighted.
Thanks very much for this change @borax12 
Merging. :fireworks:

borax12 commented:
cool @abose - looking forward to contributing more :)

abose commented:
@borax12  If you are interested in contributing to new features in brackets, please take a look at our [trello board](https://trello.com/b/LCDud1Nd/brackets) to see features requested by the community.
If you are interested in fixing something, Many unit tests are failing, which needs to be fixed :sweat_smile:
